cmd/virtiofs-placeholder/main.c

```diff
@@ -0,0 +1,193 @@
+    ltime=time(NULL); /* get current cal time */
+
+    char time_fmt[FMT_SZ] = {};
+    strftime(time_fmt, FMT_SZ, "%b %d %H:%M:%S ", localtime(&ltime));
+    fprintf(stderr, "%s", time_fmt);
+
+    fprintf(stderr, "error: ");
+    va_start(arglist, format);
+    vfprintf(stderr, format, arglist);
+    va_end(arglist);
+}
+
+int get_signalfd(int signal)
+{
+    sigset_t sigset;
+    sigemptyset(&sigset);
+    sigaddset(&sigset, signal);
+    if (sigprocmask(SIG_BLOCK, &sigset, NULL) == -1) {
+        error_log("sigprocmask failed: %s", strerror(errno));
+        return -1;
+    }
+
+    int fd = signalfd(-1, &sigset, SFD_NONBLOCK);
+    if (fd < 0) {
+        error_log("signalfd failed: %s", strerror(errno));
+    }
+    return fd;
+}
+
+int create_socket(const char *path)
+{
+    int fd = socket(AF_UNIX, SOCK_STREAM, 0);
+
+    if (fd < 0) goto err;
+
+    struct sockaddr_un addr = {};
+    addr.sun_family = AF_UNIX;
+    strncpy(addr.sun_path, path, sizeof(addr.sun_path) - 1);
+
+    if (bind(fd, (struct sockaddr *) &addr, sizeof(addr)) < 0) goto err_close;
+
+    if (listen(fd, 1) < 0) goto err_close;
+
+    return fd;
+err_close:
+    close(fd);
+err:
+    error_log("create_socket failed: %s", strerror(errno));
+    return -1;
+}
+
+int monitor(int socket_fd, int sig_fd)
+{
+    int efd = epoll_create1(0);
+    if (efd < 0) goto err;
+
+    // Watch the socket
+    // Even if we expect just one connection, we cannot use EPOLLONESHOT, because the dispatcher
+    // could have died after connect() but before spawning virtiofsd, so we need to allow successive
+    // connections.
+    struct epoll_event socket_event = {.events = EPOLLIN, .data.fd = socket_fd};
+    if (epoll_ctl(efd, EPOLL_CTL_ADD, socket_fd, &socket_event) < 0) goto err;
+
+    struct epoll_event signal_event = {.events = EPOLLIN, .data.fd = sig_fd};
+    if (epoll_ctl(efd, EPOLL_CTL_ADD, sig_fd, &signal_event) < 0) goto err;
+
+    struct epoll_event epoll_events;
+    while (true) {
+        int ret = epoll_wait(efd, &epoll_events, 1, -1);
+        if (ret < 0 && errno == EINTR) continue;
+        if (ret < 0) goto err;
+
+        if (epoll_events.data.fd == sig_fd) {
+            // We received a SIGCHLD if virtiofsd exited, we must exit too
+            struct signalfd_siginfo sfdi;
+            int len = read(epoll_events.data.fd, &sfdi, sizeof(sfdi));
+            // let's assume that only virtiofsd will run with privileges (i.e., uid == 0)
+            if (len == sizeof(sfdi) && sfdi.ssi_uid == 0) break;
+        } else if (epoll_events.data.fd == socket_fd) {
+            int accept_fd = accept(socket_fd, NULL, NULL);
+            if (accept_fd < 0) goto err;
+
+            // Get a notification if the socket is closed, to avoid leaking the FD
+            struct epoll_event acceptfd_event = {.events = EPOLLRDHUP | EPOLLONESHOT, .data.fd = accept_fd};
+            // Ignore the error, If epoll_ctl fails we will just leak the accept_fd
```

Comment:
But should we log this error? I mean do we care to monitor that `accept_fd` or not? Because if `epoll_ctl()` fails, we will not monitor that `accept_fd`, is that a problem?

Reply:
Not really, it just means we will keep the FD open, we do nothing with that FD, since the whole purpose of the socket is to allow the other side to get the PID of the placeholder

Reply:
But logging the error it is a good idea tho

Reply:
Okay, so what is the point to monitor it in that case?

I mean if we don't care about monitoring it successfully, why we are even trying to monitor it?

Reply:
Okay, I saw the comment, I think it's fine ;)